pkg/controller/cronjob/cronjob_controllerv2.go

```diff
@@ -0,0 +1,640 @@
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/component-base/metrics/prometheus/ratelimiter"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/controller"
+)
+
+var (
+	nextScheduleDelta = 100 * time.Millisecond
+)
+
+// Refactored Cronjob controller that uses DelayingQueue and informers
+
+// ControllerV2 is a controller for CronJobs.
+type ControllerV2 struct {
+	queue    workqueue.DelayingInterface
+	recorder record.EventRecorder
+
+	jobControl     jobControlInterface
+	cronJobControl cjControlInterface
+
+	jobLister     batchv1listers.JobLister
+	cronJobLister batchv1beta1listers.CronJobLister
+
+	jobListerSynced     cache.InformerSynced
+	cronJobListerSynced cache.InformerSynced
+
+	// now is a function that returns current time, done to facilitate unit tests
+	now func() time.Time
+}
+
+// NewControllerV2 creates and initializes a new Controller.
+func NewControllerV2(jobInformer batchv1informers.JobInformer, cronJobsInformer batchv1beta1informers.CronJobInformer, kubeClient clientset.Interface) (*ControllerV2, error) {
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(klog.Infof)
+	eventBroadcaster.StartRecordingToSink(&covev1client.EventSinkImpl{Interface: kubeClient.CoreV1().Events("")})
+
+	if kubeClient != nil && kubeClient.CoreV1().RESTClient().GetRateLimiter() != nil {
+		if err := ratelimiter.RegisterMetricAndTrackRateLimiterUsage("cronjob_controller", kubeClient.CoreV1().RESTClient().GetRateLimiter()); err != nil {
+			return nil, err
+		}
+	}
+
+	jm := &ControllerV2{
+		queue:    workqueue.NewNamedDelayingQueue("cronjob"),
+		recorder: eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: "cronjob-controller"}),
+
+		jobControl:     realJobControl{KubeClient: kubeClient},
+		cronJobControl: &realCJControl{KubeClient: kubeClient},
+
+		jobLister:     jobInformer.Lister(),
+		cronJobLister: cronJobsInformer.Lister(),
+
+		jobListerSynced:     jobInformer.Informer().HasSynced,
+		cronJobListerSynced: cronJobsInformer.Informer().HasSynced,
+		now:                 time.Now,
+	}
+
+	jobInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    jm.addJob,
+		UpdateFunc: jm.updateJob,
+		DeleteFunc: jm.deleteJob,
+	})
+
+	cronJobsInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			jm.enqueueController(obj)
+		},
+		UpdateFunc: jm.updateCronJob,
+		DeleteFunc: func(obj interface{}) {
+			jm.enqueueController(obj)
+		},
+	})
+
+	return jm, nil
+}
+
+// Run starts the main goroutine responsible for watching and syncing jobs.
+func (jm *ControllerV2) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer jm.queue.ShutDown()
+
+	klog.Infof("Starting cronjob controller v2")
+	defer klog.Infof("Shutting down cronjob controller v2")
+
+	go jm.enqueueAllExistingCronJobs()
```

Comment:
Do we really need that?

When starting informer, you basically will get callbacks to your Add() methods run for all existing objects.

So it should be possible to remove it.
@deads2k for final confirmation

Reply:
That's what I wasn't sure about, and the controller needs to initially feed the queue with appropriate job schedules.

Reply:
This won't even have content when it runs if the lister isn't synced.  The informer will give you clean `Add` for every cronjob.  Go ahead and remove this.

Reply:
done